cli/src/cmd/cast/wallet/mod.rs

```diff
@@ -55,15 +56,32 @@ pub enum WalletSubcommands {
         wallet: Wallet,
     },
 
-    /// Sign a message.
+    /// Sign a message or typed data.
     #[clap(visible_alias = "s")]
     Sign {
-        /// The message to sign.
+        /// The message or typed data to sign.
         ///
         /// Messages starting with 0x are expected to be hex encoded,
         /// which get decoded before being signed.
+        /// The message will be prefixed with the Ethereum Signed Message header and hashed before
+        /// signing.
+        ///
+        /// Typed data can be provided as a json string or a file name.
+        /// Use --data flag to denote the message is a string of typed data.
+        /// Use --data --from-file to denote the message is a file name containing typed data.
+        /// The data will be combined and hashed using the EIP712 specification before signing.
+        /// The data should be formatted as JSON.
         message: String,
 
+        /// If provided, the message will be treated as typed data.
+        #[clap(long)]
+        data: bool,
+
+        /// If provided, the message will be treated as a file name containing typed data. Requires
+        /// --data.
+        #[clap(long, requires = "data")]
+        from_file: bool,
```

Comment:
I think this is fine, this could try if the message argument is a file, but making this explicit is reasonable.

interpreting the message arg as file could also be added to the regular command.